aws/resource_aws_secretsmanager_secret_policy.go

```diff
@@ -0,0 +1,172 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/structure"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	iamwaiter "github.com/terraform-providers/terraform-provider-aws/aws/internal/service/iam/waiter"
+)
+
+func resourceAwsSecretsManagerSecretPolicy() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsSecretsManagerSecretPolicyCreate,
+		Read:   resourceAwsSecretsManagerSecretPolicyRead,
+		Update: resourceAwsSecretsManagerSecretPolicyUpdate,
+		Delete: resourceAwsSecretsManagerSecretPolicyDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"secret_arn": {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validateArn,
+			},
+			"policy": {
+				Type:             schema.TypeString,
+				Required:         true,
+				ValidateFunc:     validation.StringIsJSON,
+				DiffSuppressFunc: suppressEquivalentAwsPolicyDiffs,
+			},
+			"block_public_policy": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceAwsSecretsManagerSecretPolicyCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).secretsmanagerconn
+
+	input := &secretsmanager.PutResourcePolicyInput{
+		ResourcePolicy: aws.String(d.Get("policy").(string)),
+		SecretId:       aws.String(d.Get("secret_arn").(string)),
+	}
+
+	if v, ok := d.GetOk("block_public_policy"); ok {
+		input.BlockPublicPolicy = aws.Bool(v.(bool))
+	}
+
+	log.Printf("[DEBUG] Setting Secrets Manager Secret resource policy; %#v", input)
+	var res *secretsmanager.PutResourcePolicyOutput
+
+	err := resource.Retry(iamwaiter.PropagationTimeout, func() *resource.RetryError {
+		var err error
+		res, err = conn.PutResourcePolicy(input)
+		if isAWSErr(err, secretsmanager.ErrCodeMalformedPolicyDocumentException,
+			"This resource policy contains an unsupported principal") {
+			return resource.RetryableError(err)
+		}
+		if err != nil {
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+	if isResourceTimeoutError(err) {
+		res, err = conn.PutResourcePolicy(input)
```

Comment:
Why is the policy being applied again if the previous attempt timed out? Is this an expected condition?

Reply:
@bill-rich this is required in the AWS Provider due to how the AWS Go SDK works, see also:

* https://github.com/terraform-providers/terraform-provider-aws/issues/7873
* https://github.com/terraform-providers/terraform-provider-aws/issues/12985